source/shared/ed25519-lc.c

```diff
@@ -0,0 +1,351 @@
+
+    /* kdf options (we dont support it now, but still need to write out 0) */
+    if (!aws_byte_buf_write_be32(&key_buf, 0)) {
+        goto on_error;
+    }
+
+    /* number of keys */
+    if (!aws_byte_buf_write_be32(&key_buf, 1)) {
+        goto on_error;
+    }
+
+    /* encoded public key */
+    const size_t pub_encoded_len = 4 /*id len*/ + 11 /* ssh-ed25519 literal */ + 4 /*key len*/ + 32 /* key */;
+    if (!aws_byte_buf_write_be32(&key_buf, pub_encoded_len) ||
+        s_ed25519_openssh_encode_public_key(key_pair, &key_buf) != AWS_OP_SUCCESS) {
+        goto on_error;
+    }
+
+    size_t priv_block_len = 4 +                          /* check1 */
+                            4 +                          /* check2 */
+                            4 + s_key_type_literal.len + /* key type string */
+                            4 + 32 +                     /* public key */
+                            4 + 64 +                     /* private key (includes public) */
+                            4 + 0;                       /* comment (0, since comment not currently supported) */
+
+    /* pad block to the next multiple of 8 */
+    size_t priv_block_padded_len = (priv_block_len + 7) & ~7;
+
+    if (!aws_byte_buf_write_be32(&key_buf, priv_block_padded_len)) {
+        goto on_error;
+    }
+
+    uint32_t check = 0;
+    if (aws_device_random_u32(&check) != AWS_OP_SUCCESS) {
+        goto on_error;
+    }
+
+    /* check (and yeah its written twice on purpose) */
+    if (!aws_byte_buf_write_be32(&key_buf, check) || !aws_byte_buf_write_be32(&key_buf, check)) {
+        goto on_error;
+    }
+
+    /* key type */
+    if (!aws_byte_buf_write_be32(&key_buf, s_key_type_literal.len) ||
+        aws_byte_buf_append(&key_buf, &s_key_type_literal) != AWS_OP_SUCCESS) {
+        goto on_error;
+    }
+
+    /* public key (raw) */
+    if (!aws_byte_buf_write_be32(&key_buf, 32) ||
+        aws_ed25519_key_pair_get_public_key(key_pair, AWS_CAL_ED25519_KEY_EXPORT_RAW, &key_buf) != AWS_OP_SUCCESS) {
+        goto on_error;
+    }
+
+    /* private key (raw) */
+    if (!aws_byte_buf_write_be32(&key_buf, 64) ||
+        aws_ed25519_key_pair_get_private_key(key_pair, AWS_CAL_ED25519_KEY_EXPORT_RAW, &key_buf) != AWS_OP_SUCCESS) {
+        goto on_error;
+    }
+
+    /* comment */
+    if (!aws_byte_buf_write_be32(&key_buf, 0)) {
+        goto on_error;
+    }
+
+    /* padding */
+    for (uint8_t i = 1; i < (priv_block_padded_len - priv_block_len + 1); ++i) {
+        if (!aws_byte_buf_write_u8(&key_buf, i)) {
+            goto on_error;
+        }
+    }
+
+    struct aws_byte_cursor key_cur = aws_byte_cursor_from_buf(&key_buf);
+
+    if (aws_base64_encode(&key_cur, out) != AWS_OP_SUCCESS) {
+        aws_byte_buf_clean_up(&key_buf);
+        return AWS_OP_ERR;
+    }
+
+    aws_byte_buf_clean_up(&key_buf);
+    return AWS_OP_SUCCESS;
+
+on_error:
+    aws_byte_buf_clean_up(&key_buf);
+    return aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
```

Comment:
it would be nice to raise specific errors: SHORT_BUFFER vs AWS_ERROR_RANDOM_GEN_FAILED vs AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED, instead of this blanket one

Reply:
changed a bunch of places where we can definitely say why it failed. the rest are generic crypto op failed.